File: internal/application/orchestrator.go
package application

import (
	"context"
	"log"
	"sync"
	"time"

	"video-orchestrator/internal/domain"

	"github.com/sirupsen/logrus"
)

type Repository interface {
	LockAndMarkProcessing(ctx context.Context, limit int32) ([]domain.VideoJob, error)
	MarkDone(ctx context.Context, id int64) error
	MarkRetry(ctx context.Context, id int64, delay int) error
	MarkFailed(ctx context.Context, id int64) error
}

type Handler interface {
	Handle(ctx context.Context, job domain.VideoJob) error
}

type Orchestrator struct {
	repo    Repository
	handler Handler
	workers int
	breaker *CircuitBreaker
	log     *logrus.Logger
}

func NewOrchestrator(repo Repository, handler Handler, workers int, log *logrus.Logger) *Orchestrator {
	return &Orchestrator{
		repo:    repo,
		handler: handler,
		workers: workers,
		breaker: NewCircuitBreaker(5, 10*time.Second),
		log:     log,
	}
}

func (o *Orchestrator) handleFailure(ctx context.Context, job domain.VideoJob) {

	maxRetries := 5

	if job.RetryCount+1 >= maxRetries {
		if err := o.repo.MarkFailed(ctx, job.ID); err != nil {
			log.Println("failed to mark failed:", err)
		}
		return
	}

	delay := domain.NextDelay(job.RetryCount)

	o.log.WithFields(logrus.Fields{
		"job_id": job.ID,
		"retry":  job.RetryCount,
		"delay":  delay,
	}).Warn("job_retry")

	if err := o.repo.MarkRetry(ctx, job.ID, delay); err != nil {
		o.log.WithField("job_id", job.ID).
			Error("job_failed_permanently")
		log.Println("failed to mark retry:", err)
	}
}

func (o *Orchestrator) RunOnce(ctx context.Context, limit int32) {

	jobs, err := o.repo.LockAndMarkProcessing(ctx, limit)
	o.log.WithField("jobs_count", len(jobs)).Info("jobs_locked")
	if err != nil {
		log.Println(`{"event":"lock_failed"}`)
		return
	}

	var wg sync.WaitGroup
	sem := make(chan struct{}, o.workers) // worker pool

	for _, job := range jobs {
		sem <- struct{}{}
		wg.Add(1)

		go func(j domain.VideoJob) {
			o.log.WithField("job_id", j.ID).Info("job_started")
			defer wg.Done()
			defer func() { <-sem }()

			start := time.Now()

			// circuit breaker
			if !o.breaker.Allow() {
				o.log.Warn("circuit_breaker_open")
				log.Println(`{"event":"breaker_open"}`)
				return
			}

			err := o.handler.Handle(ctx, j)

			if err != nil {
				o.log.WithError(err).
					WithField("job_id", j.ID).
					Error("job_failed")
				o.breaker.Fail()
				o.handleFailure(ctx, j)
				return
			}

			o.breaker.Success()

			if err := o.repo.MarkDone(ctx, j.ID); err != nil {
				log.Println("mark_done_error:", err)
			}

			o.log.WithFields(logrus.Fields{
				"job_id": j.ID,
			}).Info("job_done")
			log.Printf(`{"event":"job_done","job_id":%d,"duration_ms":%d}`,
				j.ID,
				time.Since(start).Milliseconds(),
			)

		}(job)
	}

	wg.Wait()
}

File: application/orchestrator.ts
import type { Repository } from "../domain/repository.js";
import type { VideoJob } from "../domain/job.js";
import { canRetry } from "../domain/job.js";
import { nextDelay } from "../domain/retry.js";
import { createCircuitBreaker, type CircuitBreaker } from "./circuit_breaker.js";
import type { Logger } from "../infrastructure/logger";

export interface Handler {
    handle(job: VideoJob): Promise<void>;
}

export interface Orchestrator {
    runOnce(log: Logger): Promise<void>;
}

async function handleFailure(
    repo: Repository,
    job: VideoJob,
    maxRetries: number,
    log: Logger,
): Promise<void> {
    if (!canRetry(job, maxRetries)) {
        try {
            await repo.markFailed(job.id);
        } catch (err) {
            log.error({ err, job_id: job.id }, "mark_failed_error");
        }
        return;
    }

    const delay = nextDelay(job.retryCount);

    log.warn({ job_id: job.id, retry: job.retryCount, delay }, "job_retry");

    try {
        await repo.markRetry(job.id, delay);
    } catch (err) {
        log.error({ err, job_id: job.id }, "mark_retry_error");
    }
}

export function createOrchestrator(
    repo: Repository,
    handler: Handler,
    workers: number,
    maxRetries: number,
): Orchestrator {
    const breaker: CircuitBreaker = createCircuitBreaker(5, 10_000);

    return {
        async runOnce(log: Logger): Promise<void> {
            let jobs: VideoJob[];

            try {
                jobs = await repo.lockAndMarkProcessing(workers);
            } catch (err) {
                log.error({ err }, "lock_failed");
                return;
            }

            log.info({ jobs_count: jobs.length }, "jobs_locked");

            await Promise.all(
                jobs.map(async (job) => {
                    log.info({ job_id: job.id }, "job_started");
                    const start = Date.now();

                    if (!breaker.allow()) {
                        log.warn("circuit_breaker_open");
                        await handleFailure(repo, job, maxRetries, log);
                        return;
                    }

                    try {
                        await handler.handle(job);
                    } catch (err) {
                        log.error({ err, job_id: job.id }, "job_failed");
                        breaker.fail();
                        await handleFailure(repo, job, maxRetries, log);
                        return;
                    }

                    breaker.success();

                    try {
                        await repo.markDone(job.id);
                    } catch (err) {
                        log.error({ err, job_id: job.id }, "mark_done_error");
                    }

                    log.info({ job_id: job.id, duration_ms: Date.now() - start }, "job_done");
                }),
            );
        },
    };
}